src/java/org/apache/cassandra/net/Verb.java

```diff
@@ -434,12 +459,80 @@ public static Verb fromId(int id)
     }
 
     /**
-     * calculate an id for a custom verb
+     * Convert to/from relative and absolute id for a custom verb.
+     *
+     * <pre>{@code
+     *          relId = idForCustomVerb(absId)
+     *          absId = idForCustomVerb(relId).
+     * }</pre>
+     *
+     * <p>Relative ids can be used for indexing idToCustomVerbMap. Absolute ids exist to distinguish
+     * regular verbs from custom verbs in the id space.</p>
+     *
+     * @param id the relative or absolute id.
+     * @return a relative id if {@code id} is absolute, or absolute id if {@code id} is relative.
      */
     private static int idForCustomVerb(int id)
     {
         return CUSTOM_VERB_START - id;
     }
+
+    /**
+     * Add a new custom verb to the list of verbs.
+     *
+     * <p>While we could dynamically generate an {@code id} for callers, it's safer to have users
+     * explicitly control the id space since it prevents nodes with different versions disagreeing on which
+     * verb has which id, e.g. during upgrade.</p>
+     *
+     * @param name the name of the new verb.
+     * @param id the identifier for this custom verb (must be >= 0 and <= MAX_CUSTOM_VERB_ID).
+     * @param priority the priority of the new verb.
+     * @param expiration an optional timeout for this verb. @see VerbTimeouts.
+     * @param stage The stage this verb should execute in.
+     * @param serializer A method to serialize this verb
+     * @param handler A method to handle this verb when received by the network
+     * @param responseVerb The verb to respond with (optional)
+     * @return A Verb for the newly added verb.
+     */
+    public static synchronized Verb addCustomVerb(String name, int id, Priority priority, ToLongFunction<TimeUnit> expiration, Stage stage, Supplier<? extends IVersionedAsymmetricSerializer<?, ?>> serializer, Supplier<? extends IVerbHandler<?>> handler, Verb responseVerb)
```

Comment:
could we rename id to relativeId or document that it's relative, I took me a while to figure it out :)

Reply:
Yeah, that's fair. Fixed.